### PR TITLE
Update to libddwaf 1.5.1

### PR DESCRIPTION
**Pull request by lloeki**
**What does this PR do?**

Update libddwaf to 1.5.1

**Motivation**

Support for new features, some fixes, and performance improvements.

**How to test the change?**

There should be specs + system tests covering these.

### CODE REVIEW

File: spec/datadog/appsec/contrib/rails/integration_test_spec.rb [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+        parent = spans.find { |span| span.span_id == o.last.parent_id }
+
+        # root reached (distributed tracing)
+        break o if parent.nil?
+
+        o << parent
+      end
+    end
+    sort = ->(list) { list.sort_by { |e| chain.call(e).count } }
+    sort.call(spans)
```
Comment (ivoanjo) [inline spec/datadog/appsec/contrib/rails/integration_test_spec.rb:25]: Slightly confused -- this sort lambda doesn't seem to be reused anywhere else. Is this just a really roundabout way of just `spans.sort_by { |e| chain.call(e).count }`...?

File: spec/datadog/appsec/contrib/rails/integration_test_spec.rb [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+  end
+
+  let(:rack_span) { sorted_spans.reverse.find { |x| x.name == Datadog::Tracing::Contrib::Rack::Ext::SPAN_REQUEST } }
```
Comment (ivoanjo) [inline spec/datadog/appsec/contrib/rails/integration_test_spec.rb:28]: Are there multiple rack spans? If so, I'm curious why, and if not, I believe this whole thing can be somewhat simplified :)

File: spec/datadog/appsec/contrib/rails/integration_test_spec.rb [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+      it { expect(span.status).to eq(0) }
+
+      context 'with appsec disabled' do
+        let(:appsec_enabled) { false }
+
+        it { expect(span.get_tag('http.method')).to eq('GET') }
+        it { expect(span.get_tag('http.status_code')).to eq('200') }
+        it { expect(span.status).to eq(0) }
+      end
+    end
```
Comment (ivoanjo) [inline spec/datadog/appsec/contrib/rails/integration_test_spec.rb:92]: I noticed that the rack integration tests don't have test for appsec being disabled as we do for rails here. Would it make sense to expand those tests? (Or perhaps leave a `pending`/`skip` for now?)

File: spec/datadog/appsec/contrib/sinatra/integration_test_spec.rb [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+end
+
+require 'datadog/tracing'
+require 'datadog/tracing/metadata/ext'
+require 'datadog/tracing/sampling/ext'
+require 'datadog/tracing/contrib/sinatra/ext'
+require 'datadog/tracing/contrib/sinatra/tracer'
+
+require 'datadog/appsec'
+require 'datadog/appsec/contrib/rack/request_middleware'
```
Comment (ivoanjo) [inline spec/datadog/appsec/contrib/sinatra/integration_test_spec.rb:23]: Minor: Do we really need all of these requires? Don't `datadog/tracing` and `datadog/appsec` pull the others?

File: spec/datadog/appsec/contrib/sinatra/integration_test_spec.rb [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+      # TODO: test with c.appsec.instrument :rack
+    end
+  end
+
+  around do |example|
+    # Reset before and after each example; don't allow global state to linger.
+    Datadog.registry[:sinatra].reset_configuration!
+    example.run
+    Datadog.registry[:sinatra].reset_configuration!
+  end
```
Comment (ivoanjo) [inline spec/datadog/appsec/contrib/sinatra/integration_test_spec.rb:70]: I'm curious about the pattern of `reset_configuration!` usage in all three integration tests: rack does it only after the test, rails never resets, and sinatra does it before AND after the test.

This makes me suspect that some or all of these are either not needed, or missing. Can you doublecheck?

**Review by ivoanjo — APPROVED** [review]
I've left a few comments.

Overall seems OK, but I would strongly suggest avoiding this level of extreme complex PRs next time -- this was extremely hard to follow and review, and took a very long time :(

File: lib/datadog/appsec/assets/waf_rules/recommended.json [hunk 6]
```diff
@@ -1,9 +1,34 @@
 {
   "version": "2.2",
   "metadata": {
-    "rules_version": "1.3.1"
+    "rules_version": "1.4.1"
   },
   "rules": [
```
[the remark this answers is not in the document]
Comment (lloeki) [reply lib/datadog/appsec/assets/waf_rules/recommended.json:6]: `libddwaf` being truly just an engine that is fairly oblivious to the rules (except their format itself), it's quite independent of the ruleset. Notably, vendoring the ruleset into `libddwaf-rb` creates trickiness, e.g how to bump versions of that gem.

What I did consider though is packaging these in a separate gem and leveraging semver, so that it can be updated without doing `ddtrace` releases.

File: lib/datadog/appsec/instrumentation/gateway.rb [hunk 7]
```diff
@@ -31,8 +45,8 @@ def push(name, env, &block)
           stack.call(env)
         end
 
-        def watch(name, &block)
-          @middlewares[name] << block
+        def watch(name, key, &block)
+          @middlewares[name] << Middleware.new(key, &block) unless @middlewares[name].any? { |m| m.key == key }
         end
```
[the remark this answers is not in the document]
Comment (lloeki) [reply lib/datadog/appsec/instrumentation/gateway.rb:50]: > it's kinda weird that we keep an array + the new Middleware class and then always only have one block being set for each unique name

I understand the feeling given the currently implemented scope of AppSec. This is a major design requirement of the Instrumentation Gateway though:

- There could be multiple watchers on a given `name` in the future. In fact, during one iteration of a previous PR I had some!
- Also the Instrumentation Gateway is supposed to be factored out into Core or something, for this "high level" instrumentation system to be usable by other non-appsec ddtrace components.

The core idea behind the Instrumentation Gateway is to split the low level instrumentation (monkeypatch a class via prepend or alias method chain) from the high level one ("I want to do stuff when a rack request starts"). This separation of concerns enables two different "teams" to implement each:
- one knowledgeable with low level trickiness of whatever is being monkeypatched
- one knowledgeable with what to do with that

File: lib/datadog/appsec/contrib/rack/gateway/watcher.rb [hunk 8]
```diff
@@ -26,23 +28,25 @@ def self.watch
                   trace = active_trace
                   span = active_span
 
-                  Rack::Reactive::Request.subscribe(op, waf_context) do |action, result, _block|
-                    record = [:block, :monitor].include?(action)
+                  Rack::Reactive::Request.subscribe(op, waf_context) do |result, _block|
+                    record = [:match].include?(result.status)
                     if record
```
[the remark this answers is not in the document]
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/gateway/watcher.rb:33]: Good point.

`result` comes from the `libddwaf-rb` API though, but I'll take a note to make it better in a future update (FYI some other improvements similar to this one are planned)

File: lib/datadog/appsec/contrib/rack/gateway/watcher.rb [hunk 9]
```diff
@@ -26,23 +28,25 @@ def self.watch
...
                       event = {
                         waf_result: result,
                         trace: trace,
                         span: span,
                         request: request,
-                        action: action
+                        actions: result.actions
                       }
 
+                      span.set_tag('appsec.event', 'true') if span
```
[the remark this answers is not in the document]
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/gateway/watcher.rb:43]: When Tracing is disabled.

In that case AppSec code is supposed to run, even though it cannot record information in the trace†.

With the current state as of this PR it may seem useless (since results essentially go to `/dev/null`), but this becomes more important as e.g in-app request blocking is being progressively introduced ([hint,hint!](https://github.com/DataDog/dd-trace-rb/pull/2316))

I tried hard for this case to keep working, as it's tricky to reintroduce after the fact.

† That may change if (when?) "traces" become some sort of generic data bus upon which components such as Tracing and AppSec are built.

File: lib/datadog/appsec/contrib/rack/reactive/request.rb [hunk 10]
```diff
@@ -54,27 +54,27 @@ def self.subscribe(op, waf_context)
...
+                when :ok
                   Datadog.logger.debug { "WAF OK: #{result.inspect}" }
                 when :invalid_call
                   Datadog.logger.debug { "WAF CALL ERROR: #{result.inspect}" }
                 when :invalid_rule, :invalid_flow, :no_rule
                   Datadog.logger.debug { "WAF RULE ERROR: #{result.inspect}" }
                 else
-                  Datadog.logger.debug { "WAF UNKNOWN: #{action.inspect} #{result.inspect}" }
+                  Datadog.logger.debug { "WAF UNKNOWN: #{result.status.inspect} #{result.inspect}" }
                 end
```
[the remark this answers is not in the document]
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/reactive/request.rb:78]: It is redundant, although consider: the log message is admittedly lousy though, the idea is to immediately direct as to what is "unknown" here.

File: lib/datadog/appsec/assets/waf_rules/recommended.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (ivoanjo) [reply lib/datadog/appsec/assets/waf_rules/recommended.json:6]: > libddwaf being truly just an engine that is fairly oblivious to the rules

Yes and no. In the strictest sense, yes, someone could be using libddwaf-rb for something else other than dd-trace-rb. But... how often would that be the case? I would bet very strongly that would never, ever happen. And if it did, well, customers could open a ticket and we could decide to change our minds ;)

> Notably, vendoring the ruleset into libddwaf-rb creates trickiness, e.g how to bump versions of that gem.

Well, doesn't libddwaf-rb versioning already consider these things? E.g., if we update the rules, we release `1.5.1.0.1`, then `1.5.1.0.2`...

But! This was just a suggestion, so if you're not a big fan, we can discuss it later/afterwards/never ;)

File: lib/datadog/appsec/contrib/rack/reactive/request.rb [hunk 11]
```diff
@@ -54,27 +54,27 @@ def self.subscribe(op, waf_context)
...
+                when :match
                   Datadog.logger.debug { "WAF: #{result.inspect}" }
-                  yield [action, result, false]
-                when :block
-                  Datadog.logger.debug { "WAF: #{result.inspect}" }
-                  yield [action, result, true]
-                  throw(:block, [action, result, true])
-                when :good
+
+                  block = result.actions.include?('block')
```
[the remark this answers is not in the document]
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/reactive/request.rb:65]: `result.actions` come from [a rule's `on_match`](https://github.com/DataDog/appsec-event-rules/blob/c54492973beb80ee9c97761b80cd7768939f0a6b/build/recommended.json#L28-L30).

They're different from some other `:block` you may see here and there. That said two steps I plan to take:

- use constants to make it clear which `:block` is what (catch/throw code, return values, etc...)
- process `result.actions` dynamically, with the creation of some action handlers.

File: lib/datadog/appsec/instrumentation/gateway.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (ivoanjo) [reply lib/datadog/appsec/instrumentation/gateway.rb:50]: Fair enough! I will hold you to these words: I want to see this in core, and you'll owe me a slice of pizza if in 1 or 2 years this still is only used for appsec and with a single key :P

[the remark this answers is not in the document]
Comment (ivoanjo) [reply lib/datadog/appsec/instrumentation/gateway.rb:50]: Joking aside, it seems reasonable to try to build this in such a way if you fully expect it to become needed in the future, but I find that sometimes the crystal ball fails us (me too!) and we end up with something complex and more expensive (even in terms of memory!) that could've been simpler :)

File: Rakefile [hunk 12]
```diff
@@ -356,6 +368,18 @@ task :ci do
...
   declare '❌ 2.1 / ❌ 2.2 / ❌ 2.3 / ✅ 2.4 / ✅ 2.5 / ✅ 2.6 / ✅ 2.7 / ✅ 3.0 / ✅ 3.1 / ✅ 3.2 / ✅ jruby' => 'bundle exec appraisal cucumber4 rake spec:cucumber'
   declare '❌ 2.1 / ❌ 2.2 / ❌ 2.3 / ❌ 2.4 / ✅ 2.5 / ✅ 2.6 / ✅ 2.7 / ✅ 3.0 / ✅ 3.1 / ✅ 3.2 / ✅ jruby' => 'bundle exec appraisal cucumber5 rake spec:cucumber'
+
+  # AppSec contrib specs
+  declare '✅ 2.1 / ✅ 2.2 / ✅ 2.3 / ✅ 2.4 / ✅ 2.5 / ✅ 2.6 / ✅ 2.7 / ✅ 3.0 / ✅ 3.1 / ✅ 3.2 / ✅ jruby' => 'bundle exec appraisal contrib rake spec:appsec:rack'
+  declare '✅ 2.1 / ✅ 2.2 / ✅ 2.3 / ✅ 2.4 / ✅ 2.5 / ✅ 2.6 / ✅ 2.7 / ✅ 3.0 / ✅ 3.1 / ✅ 3.2 / ✅ jruby' => 'bundle exec appraisal contrib rake spec:appsec:sinatra'
+  declare '✅ 2.1 / ✅ 2.2 / ✅ 2.3 / ❌ 2.4 / ❌ 2.5 / ❌ 2.6 / ❌ 2.7 / ❌ 3.0 / ❌ 3.1 / ❌ 3.2 / ✅ jruby' => 'bundle exec appraisal rails4-mysql2 rake spec:rails'
+
+  # AppSec Rails specs
+  declare '✅ 2.1 / ✅ 2.2 / ✅ 2.3 / ❌ 2.4 / ❌ 2.5 / ❌ 2.6 / ❌ 2.7 / ❌ 3.0 / ❌ 3.1 / ❌ 3.2 / ✅ jruby' => 'bundle exec appraisal rails32-mysql2 rake spec:rails'
```
[the remark this answers is not in the document]
Comment (lloeki) [reply Rakefile:378]: I don't think so. As a data point I seem to recall there were Sqreen customers on Rails 3.2 + Ruby 2.1 😭 (might have been 2.0 even).

The irony is that since AppSec is a security product that could monitor and even actively protect an app, there's much incentive to add it to old apps...

File: spec/datadog/tracing/contrib/rails/support/rails3.rb [hunk 13]
```diff
@@ -99,7 +99,11 @@ def draw_test_routes!(mapper)
     test_routes = routes
     mapper.instance_exec do
       test_routes.each do |k, v|
-        get k => v
+        if k.is_a?(Array)
+          send(k.first, k.last => v)
+        else
```
[the remark this answers is not in the document]
Comment (lloeki) [reply spec/datadog/tracing/contrib/rails/support/rails3.rb:104]: Interesting, I'll try. I struggled to find something nice that would help in supporting more advanced route arguments yet not break existing code.

File: spec/datadog/appsec/processor_spec.rb [hunk 14]
```diff
@@ -356,8 +410,28 @@
 
             it { expect(matches).to have_attributes(count: 2) }
             it { expect(data).to have_attributes(count: 2) }
+            it { expect(actions).to eq [[], []] }
```
[the remark this answers is not in the document]
Comment (lloeki) [reply spec/datadog/appsec/processor_spec.rb:413]: On `'different attacks'`: when two attacks are detected in the same context there are two match results, thus two action lists, one for each.

On `'same attack'`: when the same attack is detected twice in the same context, it's only matching once (I think the rule evaluation is skipped even), therefore there's only one match result, thus one action list returned.

File: spec/datadog/appsec/processor_spec.rb [hunk 15]
```diff
@@ -356,8 +410,28 @@
...
         end
+
+        context 'one blockable attack' do
+          let(:input) { input_client_ip }
+          let(:rule_data) { rule_data_client_ip }
+
+          it { expect(matches).to have_attributes(count: 1) }
+          it { expect(data).to have_attributes(count: 1) }
+          it { expect(actions).to eq [['block']] }
+        end
```
[the remark this answers is not in the document]
Comment (lloeki) [reply spec/datadog/appsec/processor_spec.rb:424]: Yeah, notably the rule id that matched. The interesting data bits to assert against are deeply nested in a complex Hash+Array hierarchy though, which makes it hard to get to in any readable way. That's an improvement I'm planning to make by adding helpers/matchers to that end.

File: spec/datadog/appsec/contrib/rails/integration_test_spec.rb (hunk 2 above)
Comment (lloeki) [reply spec/datadog/appsec/contrib/rails/integration_test_spec.rb:28]: Yes, on nested Rack apps, which I sadly have not added the case for (yet!).

File: spec/datadog/appsec/contrib/rails/integration_test_spec.rb (hunk 3 above)
Comment (lloeki) [reply spec/datadog/appsec/contrib/rails/integration_test_spec.rb:92]: Good catch, it should be tested as well.

File: spec/datadog/appsec/contrib/sinatra/integration_test_spec.rb (hunk 5 above)
Comment (lloeki) [reply spec/datadog/appsec/contrib/sinatra/integration_test_spec.rb:70]: I took these from `spec/datadog/tracing/contrib/{rack,sinatra,rails}/integration_test_spec.rb`.

> rails never resets

I think it does (after), see:

- `spec/datadog/tracing/contrib/rails/rails_helper.rb` (especially the big final comment)
- `spec/datadog/tracing/contrib/rails/support/application.rb`

> sinatra does it before

I think this is to eagerly guard against other specs not resetting when they should (which somewhat reduces flakiness, but also makes it less likely to catch the root cause for non-after-resetting specs)

File: spec/datadog/appsec/contrib/sinatra/integration_test_spec.rb (hunk 4 above)
Comment (lloeki) [reply spec/datadog/appsec/contrib/sinatra/integration_test_spec.rb:23]: Makes sense for integration test specs.

File: spec/datadog/appsec/contrib/rack/integration_test_spec.rb [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+        context 'with an event-triggering request in application/x-www-form-url-encoded body' do
+          let(:params) { { q: '1 OR 1;' } }
+
+          let(:middlewares) do
+            [
+              Datadog::Tracing::Contrib::Rack::TraceMiddleware,
+              Datadog::AppSec::Contrib::Rack::RequestMiddleware,
+              Datadog::AppSec::Contrib::Rack::RequestBodyMiddleware,
+            ]
+          end
```
[the remark this answers is not in the document]
Comment (lloeki) [reply spec/datadog/appsec/contrib/rack/integration_test_spec.rb:223]: I'm not sure I understand the exact scope of your question 😅 

`Datadog::AppSec::Contrib::Rack::RequestBodyMiddleware` is the one needed to handle parameters coming from a request body. Note that `request.env['rack.request.form_hash']` is lazily filled and empty until `request.POST` is called, so that middleware ends up helping to ensure that.

In a real app it may need to be placed in an arbitrary location, whereas `Datadog::Tracing::Contrib::Rack::TraceMiddleware` and `Datadog::AppSec::Contrib::Rack::RequestMiddleware` should be placed near or at the top.

It is inconsequential to other test cases, and it also documents how to use these in the field, so I opted not to have a kitchensink "one size fits all" app that would add this middleware always (this use case would be best served by apps in `integration/apps`)

File: spec/datadog/appsec/contrib/rails/integration_test_spec.rb (hunk 1 above)
Comment (lloeki) [reply spec/datadog/appsec/contrib/rails/integration_test_spec.rb:25]: I think I should extract that away, I used it a lot in`spec/datadog/tracing/contrib/sinatra/tracer_spec.rb` and copied it over (BAD!)

File: lib/datadog/appsec/contrib/rack/gateway/watcher.rb [hunk 17]
```diff
@@ -137,6 +145,8 @@ def self.watch
                 [ret, res]
               end
             end
+            # rubocop:enable Metrics/CyclomaticComplexity
+            # rubocop:enable Metrics/PerceivedComplexity
             # rubocop:enable Metrics/MethodLength
             # rubocop:enable Metrics/AbcSize
```
[the remark this answers is not in the document]
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/gateway/watcher.rb:151]: Fair enough. I want to break that apart but I fear about refactoring too much things in an already big PR.

File: lib/datadog/appsec/contrib/rack/reactive/request_body.rb [hunk 18]
```diff
@@ -32,27 +32,27 @@ def self.subscribe(op, waf_context)
...
+                when :ok
                   Datadog.logger.debug { "WAF OK: #{result.inspect}" }
                 when :invalid_call
                   Datadog.logger.debug { "WAF CALL ERROR: #{result.inspect}" }
                 when :invalid_rule, :invalid_flow, :no_rule
                   Datadog.logger.debug { "WAF RULE ERROR: #{result.inspect}" }
                 else
-                  Datadog.logger.debug { "WAF UNKNOWN: #{action.inspect} #{result.inspect}" }
+                  Datadog.logger.debug { "WAF UNKNOWN: #{result.status.inspect} #{result.inspect}" }
                 end
```
[the remark this answers is not in the document]
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/reactive/request_body.rb:56]: A million times agreed.

For the longest time I kept them separate because I could not confirm if the repeating pattern would stick through more integrations, so abstracting it away would have forced the wrong abstraction.

Again, I did not want to push too much refactoring in this PR, so as stay focused on the actual goal of updating libddwaf.

File: spec/datadog/appsec/contrib/rack/integration_test_spec.rb [hunk 19]
```diff
@@ -0,0 +1,284 @@
...
+rescue LoadError
+  # fallback for old rack-contrib
+  require 'rack/contrib/post_body_content_type_parser'
+end
+
+require 'ddtrace'
+require 'datadog/tracing/contrib/rack/middlewares'
+
+require 'datadog/appsec'
+require 'datadog/appsec/contrib/rack/request_middleware'
```
[the remark this answers is not in the document]
Comment (lloeki) [reply spec/datadog/appsec/contrib/rack/integration_test_spec.rb:20]: Fair for an integration test.

File: spec/datadog/appsec/processor_spec.rb (hunk 14 above)
[the remark this answers is not in the document]
Comment (ivoanjo) [reply spec/datadog/appsec/processor_spec.rb:413]: I see. I guess you answered [below](https://github.com/DataDog/dd-trace-rb/pull/2306#discussion_r995652710) on why we don't assert further for the moment.

Until we can write code for the nicer matches, what do you think of adding the...

> On 'different attacks': when two attacks are detected in the same context there are two match results, thus two action lists, one for each.
>
> On 'same attack': when the same attack is detected twice in the same context, it's only matching once (I think the rule evaluation is skipped even), therefore there's only one match result, thus one action list returned.

...as a comment in the tests perhaps?

File: spec/datadog/appsec/contrib/rails/integration_test_spec.rb (hunk 2 above)
Comment (ivoanjo) [reply spec/datadog/appsec/contrib/rails/integration_test_spec.rb:28]: Uhh. That seems intuitively a bit weird... e.g. I guess it's an artifact of how we come in as a middleware, but as a user I'd still expect to have one single rack span even with nested apps.

(But I guess nothing to do for this PR)

File: spec/datadog/appsec/contrib/rack/integration_test_spec.rb (hunk 16 above)
[the remark this answers is not in the document]
Comment (ivoanjo) [reply spec/datadog/appsec/contrib/rack/integration_test_spec.rb:223]: My question is, being this an integration spec, it seems a bit weird to manually be modifying the middlewares. Is this something that customers will need to do to their own apps? If so, I guess the spec is realistic. But if not, can we use the same mechanism here, perhaps?

File: lib/datadog/appsec/assets/waf_rules/recommended.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (ivoanjo) [reply lib/datadog/appsec/assets/waf_rules/recommended.json:6]: > rulesets and libddwaf are extremely independent

I... somewhat disagree. Or at least, at the level we're talking about. At the level of "these are two black-box components that appsec needs to work", they are actually quite coupled. Either is quite useless without the other for 99.999999999% of our customers.

The way I see it, they are both conceptually a combined "lib-this-is-needed-for-appsec-to-actually-work".

Having rulesets inside libbdwaf I believe would enable faster releases -- because we wouldn't need to do the full QA for dd-trace-rb to a release.

I agree that all my arguments would apply as well if the rulesets were in a standalone gem, but to be honest, it still seems to me that that's exposing an internal implementation seam ("we think of them as separate") to end-users, as again, I don't expect anyone to ever reasonably separate the two. It's a bit like if I was to separate different folders of the profiler into gems. Yes, they are independent components but nobody is going to use them as so 😉 😉 😉

File: spec/datadog/appsec/contrib/rails/integration_test_spec.rb (hunk 3 above)
Comment (lloeki) [reply spec/datadog/appsec/contrib/rails/integration_test_spec.rb:92]: Actually they had it on the appsec event+tag shared examples, they were just absent of the regular sanity-check tracing ones. Anyway, added for consistency.

File: spec/datadog/appsec/contrib/rack/integration_test_spec.rb (hunk 16 above)
[the remark this answers is not in the document]
Comment (lloeki) [reply spec/datadog/appsec/contrib/rack/integration_test_spec.rb:223]: > Is this something that customers will need to do to their own apps?

Yes, in the Rack case the users build their own Rack stack, and we can't divine what it is so we can't do it automatically as with Rails or Sinatra, so they need to add our middlewares manually at the appropriate place.

In the Sinatra case, if they want JSON body parsing they're going to manually add `rack-contrib`'s `JSONBodyParser`.

So in both cases, setting up the Rack middleware stack appropriately for relevant tests is representative of what users will do.

File: spec/datadog/appsec/contrib/rack/integration_test_spec.rb [hunk 20]
```diff
@@ -0,0 +1,284 @@
...
+
+      c.appsec.enabled = appsec_enabled
+      c.appsec.instrument :rack
+    end
+  end
+
+  after { Datadog.registry[:rack].reset_configuration! }
+
+  context 'for an application' do
+    # TODO: also test without Tracing: it should run without trace transport
```
[the remark this answers is not in the document]
Comment (lloeki) [reply spec/datadog/appsec/contrib/rack/integration_test_spec.rb:41]: Task tracked internally

File: spec/datadog/tracing/contrib/rails/support/rails3.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (lloeki) [reply spec/datadog/tracing/contrib/rails/support/rails3.rb:104]: Task tracked internally

File: lib/datadog/appsec/contrib/rack/reactive/request_body.rb (hunk 18 above)
[the remark this answers is not in the document]
Comment (lloeki) [reply lib/datadog/appsec/contrib/rack/reactive/request_body.rb:56]: Task tracked internally

File: lib/datadog/appsec/assets/waf_rules/recommended.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (lloeki) [reply lib/datadog/appsec/assets/waf_rules/recommended.json:6]: I can understand that under a technical angle, but from a user's perspective the ruleset is more like a configurable input to ddtrace, the latter being a black box that happens to pull in libddwaf as a dependency. In some way, the user doesn't care at all about libddwaf but cares _a lot_ about the ruleset.

Indeed the ruleset can be tweaked, tuned, downloaded, and fed via configuration to ddtrace, entirely by the user, overriding the default that we provide. In fact, the vendored ruleset is a sort of already-primed cache, provided as a works-out-of-the-box courtesy since in theory such ruleset vendoring would be unneeded if ddtace were able to fetch it from the backend directly (hint, hint, remote config).

File: lib/datadog/appsec/instrumentation/gateway.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (lloeki) [reply lib/datadog/appsec/instrumentation/gateway.rb:50]: Yup, that's why I refrained to import the [Operation Stack](https://github.com/DataDog/datadog-instrumentation-gateway-ruby/tree/main/opstack), which, for now, is completely overkill.

File: lib/datadog/appsec/assets/waf_rules/recommended.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (lloeki) [reply lib/datadog/appsec/assets/waf_rules/recommended.json:6]: (I believe we can pursue this discussion internally if you so desire :)

File: spec/datadog/appsec/contrib/rails/integration_test_spec.rb (hunk 2 above)
Comment (lloeki) [reply spec/datadog/appsec/contrib/rails/integration_test_spec.rb:28]: (we can discuss this further internally if you wish)

File: spec/datadog/appsec/processor_spec.rb [hunk 21]
```diff
@@ -112,6 +112,26 @@
...
+
+      it { expect(described_class.new.send(:load_ruleset)).to be true }
+    end
+
+    context 'when ruleset is :risky' do
+      let(:ruleset) { :risky }
+
+      before do
+        expect(Datadog::AppSec::Assets).to receive(:waf_rules).with(:risky).and_call_original.twice
+      end
```
[the remark this answers is not in the document]
Comment (lloeki) [reply spec/datadog/appsec/processor_spec.rb:130]: Ah, it's because of the way Processor initialization + these specs are written.

1. `Processor.new` calls `#load_ruleset` internally (designed as such to have either a fully working Processor instance or an exception)
2. but we then `send(:load_ruleset)` in these examples to actually unit test that method in isolation, which is very artificial.

IRL the method is only called once.

File: spec/datadog/appsec/contrib/rails/integration_test_spec.rb (hunk 2 above)
Comment (ivoanjo) [reply spec/datadog/appsec/contrib/rails/integration_test_spec.rb:30]: You make a good point. I guess part of the trade-off here is that the "rack span" is pulling double duty: e.g. if I'm using "puma", there is no "puma span", only the "rack span".

My "intuitively a bit weird" comment was from that POV -- if I'm thinking of the "rack span" as the "puma span", then even if we do the weirdest routing things, there should only be one "puma span".

But when considering the "rack span" as the span representing a stack, which can have substacks, your point makes a lot of sense.

So... I think the current behavior seems reasonable enough, we have bigger fish to fry ;)
